drivers/lxc/driver.go

```diff
@@ -0,0 +1,510 @@
+	}
+
+	opt := toLXCCreateOptions(driverConfig)
+	if err := c.Create(opt); err != nil {
+		return nil, nil, fmt.Errorf("unable to create container: %v", err)
+	}
+
+	cleanup := func() {
+		if err := c.Destroy(); err != nil {
+			d.logger.Error("failed to clean up from an error in Start", "error", err)
+		}
+	}
+
+	if err := d.configureContainerNetwork(c); err != nil {
+		cleanup()
+		return nil, nil, err
+	}
+
+	if err := d.mountVolumes(c, cfg, driverConfig); err != nil {
+		cleanup()
+		return nil, nil, err
+	}
+
+	if err := c.Start(); err != nil {
+		cleanup()
+		return nil, nil, fmt.Errorf("unable to start container: %v", err)
+	}
+
+	if err := d.setResourceLimits(c, cfg); err != nil {
+		cleanup()
+		return nil, nil, err
+	}
+
+	pid := c.InitPid()
+
+	h := &taskHandle{
+		container:  c,
+		initPid:    pid,
+		taskConfig: cfg,
+		procState:  drivers.TaskStateRunning,
+		startedAt:  time.Now().Round(time.Millisecond),
+		logger:     d.logger,
+
+		totalCpuStats:  stats.NewCpuStats(),
+		userCpuStats:   stats.NewCpuStats(),
+		systemCpuStats: stats.NewCpuStats(),
+	}
+
+	driverState := TaskState{
+		ContainerName: c.Name(),
+		TaskConfig:    cfg,
+		StartedAt:     h.startedAt,
+	}
+
+	if err := handle.SetDriverState(&driverState); err != nil {
+		d.logger.Error("failed to start task, error setting driver state", "error", err)
+		cleanup()
+		return nil, nil, fmt.Errorf("failed to set driver state: %v", err)
+	}
+
+	d.tasks.Set(cfg.ID, h)
+
+	go h.run()
+
+	return handle, nil, nil
+}
+
+func (d *Driver) WaitTask(ctx context.Context, taskID string) (<-chan *drivers.ExitResult, error) {
+	handle, ok := d.tasks.Get(taskID)
+	if !ok {
+		return nil, drivers.ErrTaskNotFound
+	}
+
+	ch := make(chan *drivers.ExitResult)
+	go d.handleWait(ctx, handle, ch)
+
+	return ch, nil
+}
+
+func (d *Driver) handleWait(ctx context.Context, handle *taskHandle, ch chan *drivers.ExitResult) {
+	defer close(ch)
+
+	ticker := time.NewTicker(2 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-d.ctx.Done():
+			return
+		case <-ticker.C:
+			s := handle.TaskStatus()
+			if s.State == drivers.TaskStateExited {
+				ch <- handle.exitResult
+			}
+		}
+	}
+}
+
+func (d *Driver) StopTask(taskID string, timeout time.Duration, signal string) error {
+	handle, ok := d.tasks.Get(taskID)
+	if !ok {
+		return drivers.ErrTaskNotFound
+	}
+
+	if err := handle.container.Shutdown(timeout); err != nil {
+		if err := handle.container.Stop(); err != nil {
+			return fmt.Errorf("executor Shutdown failed: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (d *Driver) DestroyTask(taskID string, force bool) error {
+	handle, ok := d.tasks.Get(taskID)
+	if !ok {
+		return drivers.ErrTaskNotFound
+	}
+
+	if handle.IsRunning() && !force {
+		return fmt.Errorf("cannot destroy running task")
+	}
+
+	if handle.IsRunning() {
+		if err := handle.container.Shutdown(0); err != nil {
+			handle.logger.Error("failed to destroy executor", "err", err)
```

Comment:
Another good place for a defensive comment: why is Stop called in this block in StopTask but not here?

I can't find any reasonable guidance from the old driver, so let's just comment it:

```go
//XXX Unsure if container.Stop() should be called here when porting from pre-plugin driver
```

Or something.

Also it's strange this method doesn't call `container.Destroy()`. I'm guessing this is all why containers are left around, but again you're faithfully reproducing the old driver. Probably worth yet another comment for future readers.

Reply:
Agreed not sure why, and I simply code code over here.  `Shutdown(0)` means graceful shutdown without any timeout.  I'm guessing StopTask is called due to user interaction, while DestroyTask is called for graceful termination; and StopTask has a timeout?!  I'll add a comment.